File: azkry_android/app/src/main/kotlin/com/azkry/app/features/mushaf/models/QuranModels.kt
package com.azkry.app.features.mushaf.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** One entry of the bundled surah index (`assets/quran/index.json`). */
@Serializable
data class SurahInfo(
    @SerialName("n") val number: Int,
    val name: String,
    @SerialName("en") val englishName: String,
    /** "Meccan" or "Medinan" as shipped by the source data. */
    val type: String,
    @SerialName("count") val ayahCount: Int,
    /** Madani mushaf page of the surah's first ayah. */
    val page: Int,
    val juz: Int,
) {
    val isMeccan: Boolean get() = type == "Meccan"
}

/** First ayah of each juz (`assets/quran/index.json`). */
@Serializable
data class JuzStart(
    @SerialName("n") val number: Int,
    val surah: Int,
    val ayah: Int,
    val page: Int,
)

@Serializable
data class QuranIndex(
    val surahs: List<SurahInfo>,
    val juzs: List<JuzStart>,
)

@Serializable
data class QuranAyah(
    @SerialName("n") val numberInSurah: Int,
    @SerialName("t") val text: String,
    @SerialName("p") val page: Int,
    @SerialName("j") val juz: Int,
)

/** One surah's full text (`assets/quran/surah_<n>.json`). */
@Serializable
data class SurahContent(
    @SerialName("n") val number: Int,
    val name: String,
    val ayahs: List<QuranAyah>,
)

/** The reader's last position, persisted so the mushaf home can resume. */
data class LastRead(
    val surahNumber: Int,
    val surahName: String,
    val ayahNumber: Int,
    val page: Int,
    val timestampMillis: Long,
)

@Serializable
data class QuranBookmark(
    val surahNumber: Int,
    val surahName: String,
    val page: Int,
)

/** A khatmah plan: read the whole mushaf across [totalDays] days. */
data class Khatmah(
    val startDateKey: String,
    val totalDays: Int,
) {
    companion object {
        const val TOTAL_PAGES = 604
    }
}

private const val ARABIC_INDIC_DIGITS = "٠١٢٣٤٥٦٧٨٩"

fun Int.toArabicIndicDigits(): String =
    toString().map { ch ->
        if (ch.isDigit()) ARABIC_INDIC_DIGITS[ch - '0'] else ch
    }.joinToString("")

File: azkary_ios/Azkry/Features/Mushaf/Models/QuranModels.swift
import Foundation

struct QuranIndex: Decodable, Sendable {
    let surahs: [SurahSummary]
    let juzs: [JuzSummary]
}

struct SurahSummary: Decodable, Identifiable, Hashable, Sendable {
    let number: Int
    let name: String
    let englishName: String
    let revelationType: String
    let ayahCount: Int
    let page: Int
    let juz: Int

    var id: Int { number }

    private enum CodingKeys: String, CodingKey {
        case number = "n"
        case name
        case englishName = "en"
        case revelationType = "type"
        case ayahCount = "count"
        case page
        case juz
    }
}

struct JuzSummary: Decodable, Identifiable, Hashable, Sendable {
    let number: Int
    let surahNumber: Int
    let ayahNumber: Int
    let page: Int

    var id: Int { number }

    private enum CodingKeys: String, CodingKey {
        case number = "n"
        case surahNumber = "surah"
        case ayahNumber = "ayah"
        case page
    }
}

struct QuranSurah: Decodable, Sendable {
    let number: Int
    let name: String
    let ayahs: [QuranAyah]

    private enum CodingKeys: String, CodingKey {
        case number = "n"
        case name
        case ayahs
    }
}

struct QuranAyah: Decodable, Identifiable, Hashable, Sendable {
    let number: Int
    let text: String
    let page: Int
    let juz: Int

    var id: Int { number }

    private enum CodingKeys: String, CodingKey {
        case number = "n"
        case text = "t"
        case page = "p"
        case juz = "j"
    }
}

struct SurahOpenRequest: Hashable, Sendable {
    let surahNumber: Int
    let startAyah: Int?
    let startPage: Int?

    init(surahNumber: Int, startAyah: Int? = nil, startPage: Int? = nil) {
        self.surahNumber = surahNumber
        self.startAyah = startAyah
        self.startPage = startPage
    }
}

struct QuranLastRead: Codable, Equatable, Sendable {
    let surahNumber: Int
    let surahName: String
    let ayahNumber: Int
    let page: Int
    let updatedAt: Date
}

struct QuranBookmark: Codable, Identifiable, Hashable, Sendable {
    let surahNumber: Int
    let surahName: String
    let page: Int

    var id: Int { page }
}

struct QuranKhatmah: Codable, Equatable, Sendable {
    static let totalPages = 604

    let startDateKey: String
    let totalDays: Int
}

struct MushafPageEntry: Identifiable, Hashable, Sendable {
    let page: Int
    let surahNumber: Int
    let surahName: String

    var id: Int { page }
}

struct QuranReaderPage: Identifiable, Hashable, Sendable {
    let page: Int
    let juz: Int
    let ayahs: [QuranAyah]

    var id: Int { page }
}

struct KhatmahProgress: Equatable, Sendable {
    let dayNumber: Int
    let totalDays: Int
    let targetPage: Int
    let currentPage: Int

    var fraction: Double {
        min(1, max(0, Double(currentPage) / Double(QuranKhatmah.totalPages)))
    }
}
